corehq/apps/domain/forms.py

```diff
@@ -1621,7 +1621,7 @@ def save(self, commit=True):
                 if not account_save_success:
                     return False
 
-                cancel_future_subscriptions(self.domain, datetime.date.today(), self.creating_user)
+                cancel_future_subscriptions(self.domain, datetime.date.today(), self.creating_user, hide=True)
```

Comment:
@nickpell I added a parameter to hide cancelled subscriptions, so they are actually cancelled instead of ending the day they start (otherwise, if someone confirms a subscription here and then schedules a new one, the Subscription class will throw an error about existing future subscriptions). Are there any risks I didn't think about to doing it this way?

Reply:
> if someone confirms a subscription here and then schedules a new one, the Subscription class will throw an error about existing future subscriptions

Could you post repro steps for this?  I don't think we should get this error, because future subscriptions whose start date = end date should be ignored.

Reply:
@nickpell 
On a Pro subscription, less that 30 days after it started:
    1. Downgrade the current Pro subscription to a Standard subscription (therefore scheduling a Standard subscription to start in ~30 days)
    2. Upgrade the current Pro subscription to an Advanced subscriptions (this should end the Pro subscription now and get rid of the Pro subscription entirely)
    3. You cannot update the subscription. User sees this error message:
`It appears there was an issue subscribing your project to the Advanced Software Plan. You may try resubmitting, but if that doesn't work, rest assured someone will be contacting you shortly.`

Locally, I can see this message:
`[BILLING] There was an error subscribing the domain 'projectspace3' to plan 'CommCare Advanced'. Message: The start date of Aug 07, 2018 conflicts with the subscription dates to Subscription to CommCare Standard (v1) for DOMAIN projectspace3. [Aug 31, 2018 - Aug 31, 2018].
Traceback (most recent call last):
  File "/Users/gabriellabova/Desktop/commcare-hq/corehq/apps/domain/forms.py", line 1651, in save
    no_invoice_reason='',
  File "/Users/gabriellabova/.virtualenvs/commcare-hq/lib/python2.7/site-packages/django/utils/decorators.py", line 185, in inner
    return func(*args, **kwargs)
  File "/Users/gabriellabova/Desktop/commcare-hq/corehq/apps/accounting/models.py", line 1291, in change_plan
    new_subscription.raise_conflicting_dates(new_subscription.date_start, new_subscription.date_end)
  File "/Users/gabriellabova/Desktop/commcare-hq/corehq/apps/accounting/models.py", line 1146, in raise_conflicting_dates
    (start_before_related_end and start_after_related_start)`
One of the checks in `Subscription.raise_conflicting_dates` is blocking me from overwriting the Standard subscription with the Advanced one. Should I even bother worrying about this, or assume that any user trying to change subscriptions this many times will contact someone at Dimagi?

Reply:
For the future subscription that should be canceled, we should call ```cancel_future_subscriptions```.  This function is used in the existing self-service workflow.

Reply:
Yeah, that call to `cancel_future_subscriptions` was already in there, but `Subscription.raise_conflicting_dates` still throws an error when a future subscription (even with date_start==date_end) exists when you try to create a new subscription. My question was more about whether I should hide future subscriptions in this case, or leave the original `cancel_future_subscriptions ()` call as it is.

Reply:
Gotcha.  I think we should change ```raise_conflicting_dates``` to only check subscriptions where the end date after the start date, or is null.  Here's an example: https://github.com/dimagi/commcare-hq/blob/62957c83fffe1c4081a4878f811691ea49cb4564/corehq/apps/accounting/tasks.py#L99